### PR TITLE
feat: Implement logging for Room, RoomLifecycleManager

### LINKED ISSUE #43: [CHA-RL*] Implement logging for Room Lifecycle Manager
* Dependent on https://github.com/ably-labs/ably-chat-kotlin/pull/30
* Currently, logging is unavailable in room-lifecycle-manager, we need to implement it for debugging, understanding the lifecycle etc.
* Part of https://github.com/ably-labs/ably-chat-kotlin/issues/6

┆Issue is synchronized with this [Jira Task](https://ably.atlassian.net/browse/ECO-5095) by [Unito](https://www.unito.io)

### LINKED ISSUE #82: [CHA-RC1] Implement logging for Rooms
- Currently, logging is unavailable in `rooms.kt`, we need to implement it for debugging, understanding the lifecycle etc.
- Part of #51

┆Issue is synchronized with this [Jira Task](https://ably.atlassian.net/browse/ECO-5165) by [Unito](https://www.unito.io)

### LINKED ISSUE #83: [CHA-RC2] Implement logging for Room
- Currently, logging is unavailable in `room.kt`, we need to implement it for debugging, understanding the lifecycle etc.
- Part of #45

┆Issue is synchronized with this [Jira Task](https://ably.atlassian.net/browse/ECO-5166) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by sacOO7**
Fixed #82 
Fixed #83 
Fixed #43 
1. Added private property roomsLogger for DefaultRooms class
2. Added basic logging for Rooms#get method

## Summary by CodeRabbit

- **New Features**
	- Enhanced logging mechanisms across various components, improving error handling and traceability.
	- Introduced a new extension property to join list elements with brackets.

- **Bug Fixes**
	- Improved error handling for message processing and room operations, ensuring only valid messages trigger events.

- **Documentation**
	- Updated comments and documentation for clarity, particularly regarding logging and error handling.

- **Refactor**
	- Renamed properties related to deferred operations for better clarity in room management.

- **Tests**
	- Adjusted test cases to accommodate new logging parameters and renamed properties for consistency.

### CODE REVIEW

File: chat-android/src/main/java/com/ably/chat/RoomLifecycleManager.kt [hunk 1]
```diff
@@ -183,63 +187,84 @@ internal class RoomLifecycleManager(
...
             delay(retryDurationInMs) // Let other channels get into ATTACHING state
-            // Attach successful
-            return doAttachWithRetry()
-        } catch (ex: AblyException) {
+        } catch (ex: Throwable) {
             // CHA-RL5c - Channel attach failed
-            statusLifecycle.setStatus(RoomStatus.Failed, ex.errorInfo)
+            logger.error("doRetry(); waiting failed for channel attach, feature: ${contributor.featureName}", ex)
+            // No need for channels wind down, since it's taken care at the start of doRetry()
+            statusLifecycle.setStatus(RoomStatus.Failed, (ex as? AblyException)?.errorInfo)
```
[the remark this answers is not in the document]
Comment (sacOO7) [reply chat-android/src/main/java/com/ably/chat/RoomLifecycleManager.kt:263]: We are okay even if `errorInfo` is `null`, so there's no need to create custom error if it's null.

File: chat-android/src/main/java/com/ably/chat/Rooms.kt [hunk 2]
```diff
@@ -72,49 +73,59 @@ internal class DefaultRooms(
...
+                if (roomReleaseDeferredMap.containsKey(roomId)) {
+                    roomReleaseDeferredMap[roomId]?.await()
                 } else {
                     val roomReleaseDeferred = CompletableDeferred<Unit>()
-                    this@DefaultRooms.roomReleaseDeferred[roomId] = roomReleaseDeferred
+                    roomReleaseDeferredMap[roomId] = roomReleaseDeferred
                     existingRoom.release() // CHA-RC1g5
                     roomReleaseDeferred.complete(Unit)
                 }
+                logger.debug("release(); released roomId: $roomId")
```
[the remark this answers is not in the document]
Comment (sacOO7) [reply chat-android/src/main/java/com/ably/chat/Rooms.kt:126]: All operations in `Rooms.kt` are executed under `sequentialScope` with parallelism set to `1`, so we will never face concurrency issues

File: chat-android/src/main/java/com/ably/chat/Rooms.kt [hunk 3]
```diff
@@ -127,33 +138,33 @@
     private suspend fun getReleasedOrExistingRoom(roomId: String): Room? {
         // Previous Room Get in progress, because room release in progress
         // So await on same deferred and return null
-        roomGetDeferred[roomId]?.let {
+        roomGetDeferredMap[roomId]?.let {
+            logger.debug("getReleasedOrExistingRoom(); awaiting on previous rooms.get() for roomId: $roomId")
```
[the remark this answers is not in the document]
Comment (sacOO7) [reply chat-android/src/main/java/com/ably/chat/Rooms.kt:142]: Same as https://github.com/ably/ably-chat-kotlin/pull/84/files#r1881684478

**Comment by sacOO7** [comment]
Unit tests for logging will be addressed as a part of https://github.com/ably/ably-chat-kotlin/issues/92

**Review by ttypic — APPROVED** [review]
LGTM

**Pull request merged**